# Testing MultiIndex options

For this code, i wanted to be able to move/choose index. I'm using Financial datas of Google and Microsoft.<br /><br />
- I get data as usually, and tried MultiIndex function. But because of this multi-level dataframe, i didn't reach what i wanted to.
- Then i tried using stack() and reindex() but didn't worked
- The solution i found is getting data for each companies (no multi-level dataframe) and concate dataframes. 

A function can probably do the job (and improve the code), with a list of companies we want.<br />
In the part 1.1, the transpose() function offers a satisfying result, but i wanted to check options i had.<br />
Same in part 1.2 for the stack() function.

In [1]:
# Import librairies & Set
import pandas as pd
import numpy as np
import datetime as dt
from pandas_datareader import data

pd.set_option('display.float_format', '{:20,.2f}'.format)

## First try (double companies)

In [2]:
company = ['MSFT', 'GOOG']
start = "2010-01-01"
end = "2019-08-13"

stocks = data.DataReader(name = company, data_source = "yahoo", start = start, end = end)
stocks.head()

Attributes                 High                                       Low  \
Symbols                    GOOG                 MSFT                 GOOG   
Date                                                                        
2010-01-04               313.58                31.10               310.95   
2010-01-05               312.75                31.10               309.61   
2010-01-06               311.76                31.08               302.05   
2010-01-07               303.86                30.70               295.22   
2010-01-08               300.50                30.88               293.46   

Attributes                                      Open                       \
Symbols                    MSFT                 GOOG                 MSFT   
Date                                                                        
2010-01-04                30.59               312.30                30.62   
2010-01-05                30.64               312.42                30.85   
2010-01-06                30.52               311.76                30.88   
2010-01-07                30.19               303.56                30.63   
2010-01-08                30.24               294.89                30.28   

Attributes                Close                                    Volume  \
Symbols                    GOOG                 MSFT                 GOOG   
Date                                                                        
2010-01-04               312.20                30.95         3,927,000.00   
2010-01-05               310.83                30.96         6,031,900.00   
2010-01-06               302.99                30.77         7,987,100.00   
2010-01-07               295.94                30.45        12,876,600.00   
2010-01-08               299.89                30.66         9,483,900.00   

Attributes                                 Adj Close                       
Symbols                    MSFT                 GOOG                 MSFT  
Date                                                                       
2010-01-04        38,409,100.00               312.20                24.44  
2010-01-05        49,749,600.00               310.83                24.45  
2010-01-06        58,182,400.00               302.99                24.30  
2010-01-07        50,559,700.00               295.94                24.05  
2010-01-08        51,197,400.00               299.89                24.21

In [3]:
df = stocks.transpose()
df.head()

Date                         2010-01-04           2010-01-05  \
Attributes Symbols                                             
High       GOOG                  313.58               312.75   
           MSFT                   31.10                31.10   
Low        GOOG                  310.95               309.61   
           MSFT                   30.59                30.64   
Open       GOOG                  312.30               312.42   

Date                         2010-01-06           2010-01-07  \
Attributes Symbols                                             
High       GOOG                  311.76               303.86   
           MSFT                   31.08                30.70   
Low        GOOG                  302.05               295.22   
           MSFT                   30.52                30.19   
Open       GOOG                  311.76               303.56   

Date                         2010-01-08           2010-01-11  \
Attributes Symbols                                             
High       GOOG                  300.50               301.10   
           MSFT                   30.88                30.76   
Low        GOOG                  293.46               295.91   
           MSFT                   30.24                30.12   
Open       GOOG                  294.89               301.10   

Date                         2010-01-12           2010-01-13  \
Attributes Symbols                                             
High       GOOG                  297.96               293.09   
           MSFT                   30.40                30.52   
Low        GOOG                  292.90               285.88   
           MSFT                   29.91                30.01   
Open       GOOG                  297.71               287.17   

Date                         2010-01-14           2010-01-15  ...  \
Attributes Symbols                                            ...   
High       GOOG                  295.99               295.67  ...   
           MSFT                   31.10                31.24  ...   
Low        GOOG                  290.32               287.94  ...   
           MSFT                   30.26                30.71  ...   
Open       GOOG                  290.86               295.56  ...   

Date                         2019-07-31           2019-08-01  \
Attributes Symbols                                             
High       GOOG                1,234.00             1,234.11   
           MSFT                  140.49               140.94   
Low        GOOG                1,207.76             1,205.72   
           MSFT                  135.08               136.93   
Open       GOOG                1,223.00             1,214.03   

Date                         2019-08-02           2019-08-05  \
Attributes Symbols                                             
High       GOOG                1,206.90             1,175.24   
           MSFT                  138.32               133.93   
Low        GOOG                1,188.94             1,140.14   
           MSFT                  135.26               130.78   
Open       GOOG                1,200.74             1,170.04   

Date                         2019-08-06           2019-08-07  \
Attributes Symbols                                             
High       GOOG                1,179.96             1,178.44   
           MSFT                  135.68               135.65   
Low        GOOG                1,160.00             1,149.62   
           MSFT                  133.21               131.83   
Open       GOOG                1,163.31             1,156.00   

Date                         2019-08-08           2019-08-09  \
Attributes Symbols                                             
High       GOOG                1,205.01             1,203.88   
           MSFT                  138.99               139.38   
Low        GOOG                1,173.02             1,183.60   
           MSFT                  135.93              

In [4]:
index = pd.MultiIndex.from_product([['GOOG', 'MSFT'],
                                    ['High', 'Low', 'Open', 'Close', 'Volume', 'Adj Close']], 
                                   names=['State', 'Direction'])
df = pd.DataFrame(index=index,data=df.values, columns=df.columns)
df.head()

Date                      2010-01-04           2010-01-05  \
State Direction                                             
GOOG  High                    313.58               312.75   
      Low                      31.10                31.10   
      Open                    310.95               309.61   
      Close                    30.59                30.64   
      Volume                  312.30               312.42   

Date                      2010-01-06           2010-01-07  \
State Direction                                             
GOOG  High                    311.76               303.86   
      Low                      31.08                30.70   
      Open                    302.05               295.22   
      Close                    30.52                30.19   
      Volume                  311.76               303.56   

Date                      2010-01-08           2010-01-11  \
State Direction                                             
GOOG  High                    300.50               301.10   
      Low                      30.88                30.76   
      Open                    293.46               295.91   
      Close                    30.24                30.12   
      Volume                  294.89               301.10   

Date                      2010-01-12           2010-01-13  \
State Direction                                             
GOOG  High                    297.96               293.09   
      Low                      30.40                30.52   
      Open                    292.90               285.88   
      Close                    29.91                30.01   
      Volume                  297.71               287.17   

Date                      2010-01-14           2010-01-15  ...  \
State Direction                                            ...   
GOOG  High                    295.99               295.67  ...   
      Low                      31.10                31.24  ...   
      Open                    290.32               287.94  ...   
      Close                    30.26                30.71  ...   
      Volume                  290.86               295.56  ...   

Date                      2019-07-31           2019-08-01  \
State Direction                                             
GOOG  High                  1,234.00             1,234.11   
      Low                     140.49               140.94   
      Open                  1,207.76             1,205.72   
      Close                   135.08               136.93   
      Volume                1,223.00             1,214.03   

Date                      2019-08-02           2019-08-05  \
State Direction                                             
GOOG  High                  1,206.90             1,175.24   
      Low                     138.32               133.93   
      Open                  1,188.94             1,140.14   
      Close                   135.26               130.78   
      Volume                1,200.74             1,170.04   

Date                      2019-08-06           2019-08-07  \
State Direction                                             
GOOG  High                  1,179.96             1,178.44   
      Low                     135.68               135.65   
      Open                  1,160.00             1,149.62   
      Close                   133.21               131.83   
      Volume                1,163.31             1,156.00   

Date                      2019-08-08           2019-08-09  \
State Direction                                             
GOOG  High                  1,205.01             1,203.88   
      Low                     138.99               139.38   
      Open                  1,173.02             1,183.60   
      Close                   135.93               136.46   
      Volume                1,182.83             1,197.99   

Date                      2019-08-12           2019-08-13  
State Direction                                      

Looking good, but data are wrong. <br />
<br />
Example :<br />
GOOG, Low, 2010-01-04 = 31.10<br />
This is the value for MSFT, High, 2010-01-04

## Second try (stack() + reindex())

In [5]:
stocks.stack().head()

Attributes                         High                  Low  \
Date       Symbols                                             
2010-01-04 GOOG                  313.58               310.95   
           MSFT                   31.10                30.59   
2010-01-05 GOOG                  312.75               309.61   
           MSFT                   31.10                30.64   
2010-01-06 GOOG                  311.76               302.05   

Attributes                         Open                Close  \
Date       Symbols                                             
2010-01-04 GOOG                  312.30               312.20   
           MSFT                   30.62                30.95   
2010-01-05 GOOG                  312.42               310.83   
           MSFT                   30.85                30.96   
2010-01-06 GOOG                  311.76               302.99   

Attributes                       Volume            Adj Close  
Date       Symbols                                            
2010-01-04 GOOG            3,927,000.00               312.20  
           MSFT           38,409,100.00                24.44  
2010-01-05 GOOG            6,031,900.00               310.83  
           MSFT           49,749,600.00                24.45  
2010-01-06 GOOG            7,987,100.00               302.99

In [6]:
index = pd.MultiIndex.from_product([['GOOG', 'MSFT'], 
                                    ['High', 'Low', 'Open', 'Close', 'Volume', 'Adj Close']], 
                                   names=['Symbols', 'Attributes'])
stocks.reindex(index).head()

Attributes                         High                       \
Symbols                            GOOG                 MSFT   
Symbols Attributes                                             
GOOG    High                        nan                  nan   
        Low                         nan                  nan   
        Open                        nan                  nan   
        Close                       nan                  nan   
        Volume                      nan                  nan   

Attributes                          Low                       \
Symbols                            GOOG                 MSFT   
Symbols Attributes                                             
GOOG    High                        nan                  nan   
        Low                         nan                  nan   
        Open                        nan                  nan   
        Close                       nan                  nan   
        Volume                      nan                  nan   

Attributes                         Open                       \
Symbols                            GOOG                 MSFT   
Symbols Attributes                                             
GOOG    High                        nan                  nan   
        Low                         nan                  nan   
        Open                        nan                  nan   
        Close                       nan                  nan   
        Volume                      nan                  nan   

Attributes                        Close                       \
Symbols                            GOOG                 MSFT   
Symbols Attributes                                             
GOOG    High                        nan                  nan   
        Low                         nan                  nan   
        Open                        nan                  nan   
        Close                       nan                  nan   
        Volume                      nan                  nan   

Attributes                       Volume                       \
Symbols                            GOOG                 MSFT   
Symbols Attributes                                             
GOOG    High                        nan                  nan   
        Low                         nan                  nan   
        Open                        nan                  nan   
        Close                       nan                  nan   
        Volume                      nan                  nan   

Attributes                    Adj Close                       
Symbols                            GOOG                 MSFT  
Symbols Attributes                                            
GOOG    High                        nan                  nan  
        Low                         nan                  nan  
        Open                        nan                  nan  
        Close                       nan                  nan  
        Volume                      nan                  nan

## Third try (reduce dimension, companies one by one)

In [7]:
company = 'GOOG' # without brackets
start = "2010-01-01"
end = "2019-08-13"

stocksss = data.DataReader(name = company, data_source = "yahoo", start = start, end = end)
stocksss.head()

High                  Low                 Open  \
Date                                                                        
2010-01-04               313.58               310.95               312.30   
2010-01-05               312.75               309.61               312.42   
2010-01-06               311.76               302.05               311.76   
2010-01-07               303.86               295.22               303.56   
2010-01-08               300.50               293.46               294.89   

                          Close               Volume            Adj Close  
Date                                                                       
2010-01-04               312.20         3,927,000.00               312.20  
2010-01-05               310.83         6,031,900.00               310.83  
2010-01-06               302.99         7,987,100.00               302.99  
2010-01-07               295.94        12,876,600.00               295.94  
2010-01-08               299.89         9,483,900.00               299.89

In [8]:
ml = stocksss.transpose()
index = pd.MultiIndex.from_product([['GOOG'], 
                                    ['High', 'Low', 'Open', 'Close', 'Volume', 'Adj Close']], 
                                    names=['State', 'Direction'])
df = pd.DataFrame(index=index,data=ml.values, columns=ml.columns)
df.head()

Date                      2010-01-04           2010-01-05  \
State Direction                                             
GOOG  High                    313.58               312.75   
      Low                     310.95               309.61   
      Open                    312.30               312.42   
      Close                   312.20               310.83   
      Volume            3,927,000.00         6,031,900.00   

Date                      2010-01-06           2010-01-07  \
State Direction                                             
GOOG  High                    311.76               303.86   
      Low                     302.05               295.22   
      Open                    311.76               303.56   
      Close                   302.99               295.94   
      Volume            7,987,100.00        12,876,600.00   

Date                      2010-01-08           2010-01-11  \
State Direction                                             
GOOG  High                    300.50               301.10   
      Low                     293.46               295.91   
      Open                    294.89               301.10   
      Close                   299.89               299.43   
      Volume            9,483,900.00        14,479,800.00   

Date                      2010-01-12           2010-01-13  \
State Direction                                             
GOOG  High                    297.96               293.09   
      Low                     292.90               285.88   
      Open                    297.71               287.17   
      Close                   294.14               292.45   
      Volume            9,742,900.00        13,041,800.00   

Date                      2010-01-14           2010-01-15  ...  \
State Direction                                            ...   
GOOG  High                    295.99               295.67  ...   
      Low                     290.32               287.94  ...   
      Open                    290.86               295.56  ...   
      Close                   293.82               288.92  ...   
      Volume            8,511,900.00        10,909,600.00  ...   

Date                      2019-07-31           2019-08-01  \
State Direction                                             
GOOG  High                  1,234.00             1,234.11   
      Low                   1,207.76             1,205.72   
      Open                  1,223.00             1,214.03   
      Close                 1,216.68             1,209.01   
      Volume            1,725,500.00         1,698,500.00   

Date                      2019-08-02           2019-08-05  \
State Direction                                             
GOOG  High                  1,206.90             1,175.24   
      Low                   1,188.94             1,140.14   
      Open                  1,200.74             1,170.04   
      Close                 1,193.99             1,152.32   
      Volume            1,645,100.00         2,597,500.00   

Date                      2019-08-06           2019-08-07  \
State Direction                                             
GOOG  High                  1,179.96             1,178.44   
      Low                   1,160.00             1,149.62   
      Open                  1,163.31             1,156.00   
      Close                 1,169.95             1,173.99   
      Volume            1,709,400.00         1,444,300.00   

Date                      2019-08-08           2019-08-09  \
State Direction                                             
GOOG  High                  1,205.01             1,203.88   
      Low                   1,173.02             1,183.60   
      Open                  1,182.83             1,197.99   
      Close                 1,204.80             1,188.01   
      Volume            1,468,000.00         1,065,700.00   

Date                      2019-08-12           2019-08-13  
State Direction                                      

In [9]:
company = 'MSFT'
start = "2010-01-01"
end = "2019-08-13"

stocksss = data.DataReader(name = company, data_source = "yahoo", start = start, end = end)
ml = stocksss.transpose()
index = pd.MultiIndex.from_product([['MSFT'], 
                                    ['High', 'Low', 'Open', 'Close', 'Volume', 'Adj Close']], 
                                    names=['State', 'Direction'])
dff = pd.DataFrame(index=index,data=ml.values, columns=ml.columns)
dff.head()

Date                      2010-01-04           2010-01-05  \
State Direction                                             
MSFT  High                     31.10                31.10   
      Low                      30.59                30.64   
      Open                     30.62                30.85   
      Close                    30.95                30.96   
      Volume           38,409,100.00        49,749,600.00   

Date                      2010-01-06           2010-01-07  \
State Direction                                             
MSFT  High                     31.08                30.70   
      Low                      30.52                30.19   
      Open                     30.88                30.63   
      Close                    30.77                30.45   
      Volume           58,182,400.00        50,559,700.00   

Date                      2010-01-08           2010-01-11  \
State Direction                                             
MSFT  High                     30.88                30.76   
      Low                      30.24                30.12   
      Open                     30.28                30.71   
      Close                    30.66                30.27   
      Volume           51,197,400.00        68,754,700.00   

Date                      2010-01-12           2010-01-13  \
State Direction                                             
MSFT  High                     30.40                30.52   
      Low                      29.91                30.01   
      Open                     30.15                30.26   
      Close                    30.07                30.35   
      Volume           65,912,100.00        51,863,500.00   

Date                      2010-01-14           2010-01-15  ...  \
State Direction                                            ...   
MSFT  High                     31.10                31.24  ...   
      Low                      30.26                30.71  ...   
      Open                     30.31                31.08  ...   
      Close                    30.96                30.86  ...   
      Volume           63,228,100.00        79,913,200.00  ...   

Date                      2019-07-31           2019-08-01  \
State Direction                                             
MSFT  High                    140.49               140.94   
      Low                     135.08               136.93   
      Open                    140.33               137.00   
      Close                   136.27               138.06   
      Volume           38,598,800.00        40,557,500.00   

Date                      2019-08-02           2019-08-05  \
State Direction                                             
MSFT  High                    138.32               133.93   
      Low                     135.26               130.78   
      Open                    138.09               133.30   
      Close                   136.90               132.21   
      Volume           30,791,600.00        42,749,600.00   

Date                      2019-08-06           2019-08-07  \
State Direction                                             
MSFT  High                    135.68               135.65   
      Low                     133.21               131.83   
      Open                    133.80               133.79   
      Close                   134.69               135.28   
      Volume           32,696,700.00        33,414,500.00   

Date                      2019-08-08           2019-08-09  \
State Direction                                             
MSFT  High                    138.99               139.38   
      Low                     135.93               136.46   
      Open                    136.60               138.61   
      Close                   138.89               137.71   
      Volume           27,496,500.00        23,466,700.00   

Date                      2019-08-12           2019-08-13  
State Direction                                      

## Ok, perfect. Let's concat

In [10]:
frames = [df, dff]
result = pd.concat(frames)
result

Date                      2010-01-04           2010-01-05  \
State Direction                                             
GOOG  High                    313.58               312.75   
      Low                     310.95               309.61   
      Open                    312.30               312.42   
      Close                   312.20               310.83   
      Volume            3,927,000.00         6,031,900.00   
      Adj Close               312.20               310.83   
MSFT  High                     31.10                31.10   
      Low                      30.59                30.64   
      Open                     30.62                30.85   
      Close                    30.95                30.96   
      Volume           38,409,100.00        49,749,600.00   
      Adj Close                24.44                24.45   

Date                      2010-01-06           2010-01-07  \
State Direction                                             
GOOG  High                    311.76               303.86   
      Low                     302.05               295.22   
      Open                    311.76               303.56   
      Close                   302.99               295.94   
      Volume            7,987,100.00        12,876,600.00   
      Adj Close               302.99               295.94   
MSFT  High                     31.08                30.70   
      Low                      30.52                30.19   
      Open                     30.88                30.63   
      Close                    30.77                30.45   
      Volume           58,182,400.00        50,559,700.00   
      Adj Close                24.30                24.05   

Date                      2010-01-08           2010-01-11  \
State Direction                                             
GOOG  High                    300.50               301.10   
      Low                     293.46               295.91   
      Open                    294.89               301.10   
      Close                   299.89               299.43   
      Volume            9,483,900.00        14,479,800.00   
      Adj Close               299.89               299.43   
MSFT  High                     30.88                30.76   
      Low                      30.24                30.12   
      Open                     30.28                30.71   
      Close                    30.66                30.27   
      Volume           51,197,400.00        68,754,700.00   
      Adj Close                24.21                23.91   

Date                      2010-01-12           2010-01-13  \
State Direction                                             
GOOG  High                    297.96               293.09   
      Low                     292.90               285.88   
      Open                    297.71               287.17   
      Close                   294.14               292.45   
      Volume            9,742,900.00        13,041,800.00   
      Adj Close               294.14               292.45   
MSFT  High                     30.40                30.52   
      Low                      29.91                30.01   
      Open                     30.15                30.26   
      Close                    30.07                30.35   
      Volume           65,912,100.00        51,863,500.00   
      Adj Close                23.75                23.97   

Date                      2010-01-14           2010-01-15  ...  \
State Direction                                            ...   
GOOG  High                    295.99               295.67  ...   
      Low                     290.32               287.94  ...   
      Open                    290.86               295.56  ...   
      Close                   293.82               288.92  ...   
      Volume            8,511,900.00        10,909,600.00  ...   
      Adj Close               293.82               288.92  ...   
MSFT  High                     31.10                